### ISSUE TITLE
Discuss project direction

### ISSUE THREAD

**Issue by SeanPrashad** [Mera-Gangapersaud/Creative-Collab#21]
Original comment found [here](https://github.com/Mera-Gangapersaud/Creative-Collab/issues/13#issuecomment-437635827)

After digging into the project for a while, I personally think we are going too fast, making it's hard for further development. What we are lacking of:

- The workflow: current description is too general.
  - When users first get into the app, do they see the game board? No, they need to join a room. So what do they see first? Room list? Button to join? Create room button?
  - If they create rooms, what's next? Game board now? No, it should be a form with some kind of info, player name? Max number of player? Room name?
  - Now in game board, at first the host has to wait for other players. How do they know when the game starts? How do they know if it's their turns? Should we have a kind of notification bar or pop up? Now that will affect the interface a lot
  - So on...
- The design: this I believe the orders should be: wireframes -> implementations, not the other way around. For ex we need something like this first, before starting to code:
![untitled](https://user-images.githubusercontent.com/31667324/48307835-a59dd300-e523-11e8-9c83-e2b13d5d3504.jpg)
- Now with the two drafts above, we can have a frame for the app, which is very important for collaborating works to be consistent. Our `master` branch is not well organized. The entry `CreativeCollab.jsx` is a test file? Should we use `App` component like other projects? The current `App` component is `TicTacToe` game, which is just for research purpose, why it is in `master` branch? Should we put all components in a folder? For testing, should we include the test to repository or put it in `.gitignore`? We can do something similar to [vuetify](https://github.com/vuetifyjs/vuetify), don't include the playground file in the source code, but put them in description of every PR. And so on...

In short words, I believe the app need to be well organized first. Issues like #5 or #9 are good example of making the app bones. This is the very first thing we should do before going for further developments. I'm currently working on #10, and found it quite hard to code because it might conflict with the works of other contributors.

Just some personal opinions 😋 

P/S: #20 is exactly what I wanna say

_Originally posted by @pynnl in https://github.com/Mera-Gangapersaud/Creative-Collab/issues/13#issuecomment-437635827_

**Comment by SeanPrashad** [comment]
# Follow up by @dnguneratne

I agree with @pynnl. A few points:

- I feel like the current code base has been created without fully realizing how boardgame.io framework works, which is the reason for all the changes introduced in #20.
- When I was working on #10 I considered the same issues mentioned in @pynnl's point in `The workflow`. We need to flush out the details and expectations.
- As for the design, PR #20 is a start but it is still only client side. We need to explore how boardgame.io is meant to work with a master server.

**Comment by SeanPrashad** [comment]
First and foremost - @Mera-Gangapersaud and I chatted briefly and we both love the amount of effort the both of you have invested. There's some very valid points brought up and we want to discuss them together to see how we can help get #20 landed.

**I am speaking only for myself from here on**:

Addressing @pynnl:

Again, love the wire framing that you've done - I don't think I've seen anyone else draft something like that since we were taught how to do it awhile ago. I'm able to clearly see what the end product looks for you and it's great for discussion!

I do agree that we are moving fast - this isn't by any means just a "hack it together" kind of project, I know @Mera-Gangapersaud has a part of her devoted to this project and she'd love to see it flourish (as all of us contributing would). Many of the ideas you have are prime material for opening new issues (_Create an initial landing page_, _Allow players to join rooms_, etc etc) and I definitely encourage you to actually open those issues to get discussion going - I would be happy to help mentor any of those that interest you! 🙌🏽

We wanted to get started quickly as we have ~3 weeks of class left (not to mention other course work). Mera had conducted some light research into boardgame.io and found it was suitable for what we wanted, so we got up and running fast. I envisioned carving out a community where any contributor could help (_from individuals new to React like Deepanjali to experienced devs like Dilan_) and to do so mean't that we had to get a "base" established, no matter how fragile it may be. 

But with that being said, we've moved past our "0.1" of researching frameworks to use and I agree that a bit more formality, both in our project issues and the culture of how contributors work together, is something that we should focus on. Your well though out post made me realize that I should focus more on dispersing ambiguity throughout the project and establishing "milestones" that we collectively can aim for.

Tl;DR: Let's create some more issues using that will help establish guidelines for where we should be headed.

Addressing @dnguneratne:

First off, never in my wildest dreams did I expect for a contributor to stroll along and submit a patch that fixed so much - me IRL: 🤯😳😬😧 (a huge thank you). We **will** get #20 landed, even if it needs rebasing (which I am more than happy to sit down and help walk you through).

Dave left some comments which I mutually agree with. I would have told everyone "_HEY STAY CLEAR OF THAT ISSUE - IT IS TAKEN!!!_" if you gave me a heads up that you were working on it. It's not a problem that we can't fix, but something to keep in mind for the future, especially when you start work and have to keep track of what you're assigned through JIRA or whatever.

It would also be great for you to evolve your git game and make smaller, more granular commits! I know it's something we all can do but being lazy is easy, and easy only goes so far in this world - that's my challenge to you ✌🏽

Let's leave it there for now - our project functions as a community and I want to hear from the others and what they have to say. In the meantime, I'm going to chip away at reviewing #20 and opening more issues over the next few days.

Any more suggestions, thoughts, and comments are welcome - there is no wrong answer here.

**Comment by pynnl** [comment]
I suggest these steps after #20 land:
- Decide how many interfaces we need, and put them in different route (use `react-router`), for ex: `\`, `\about`, `\room\:id`. This can be filed in an issue. After this is solved, contributors can start to work on each route without worrying about conflicts.
- For each route (interface or page, call it as you wish 😋 ), decide the workflow and layout. This should be done before we start to code anything. This can be filed in several issues, 1 issue for 1 route. The PR should setup the layout for the interface. For ex, say in `\` there are 3 sections `join`, `create room`, `room list`, the PR will create 3 components (3 more js files) and import them inside the `\` component (just a remind, each route is a component). After that we can file issues for these components (keep them small and well-organized 😄 ).
- Server side: the workflow between client and server is simple as in https://boardgame.io/#/api/Server, but there are things to be tested out:
  - GET `/game`: get room list
  - POST `/games/:game-name/:room-id/join`: join the game with specific `room-id` (we get room lists and pick from the one which are not full). However I'm confused where we must provide a `playerId` to join a room. What If 2 clients request to join with the same `playerId` at the same time? Need to test that out cause it's not mentioned in the doc.
  - POST `game/../../leave`: to leave a room, say a player quit while the game is being played. This is fine if players press the **leave button**. However what if they just close the tab or the browser? There will be no request to the server, can the server api handle that?
  - There will be no PR for these researchs, so I hope we will have volunteers for this.

**Comment by dnguneratne** [comment]
The problem I see with just doing research is the individual that did the research is the one that will be able to implement this feature. We could ask them to create branches that show their work with mini examples that others can follow.

Alternatively, we can break these down to issues that people can pick them up, research and implement. For example, an issue could be "implement rooms" or "implement ability to join a room", which then would require research on how Rooms work. It could get a little messy depending on how many parts needs to be moved around / re-implemented.

Perhaps we could use a two milestones. One for Research that includes things mentioned by @pynnl in "Server side" and more that I can add, and the other for Design or mocks. Either that or we can create issues labeled Design or Research / Experimentation.

Another thing to consider is what is the workflow when some issues are dependent on others? How about using the GitHub project board or Trello?

Lastly, @Mera-Gangapersaud maybe there's value in creating a requirements doc (wiki or updated Readme) to flush out the vision and scope of this project. For example, the initial project description doesn't call for the ability to join a room. Is that functionality out of scope? The description also says "This will build the back end of the app". PR #20 contains both back-end and front-end in separate folders. Did you mean to split them in to two different repositories? What other things need to be considered?

**Comment by Mera-Gangapersaud** [comment]
To reiterate what @SeanPrashad said; I am impressed and appreciative of the advice and work both @pynnl and @dnguneratne have put into Creative Collab so far. 

In regards to @pynnl's point about moving to fast; I've realized it's because we started on step 5 rather than step 1 and as a result development is difficult. 

To address this I've followed your lead with the wireframes and created some that we can discuss further [over in a wiki](https://github.com/Mera-Gangapersaud/Creative-Collab/wiki/WireFrame). To stick with the nature of this game I've changed some of the terminology (ie. room to story, player to writer). I'm hoping this will serve as better guidance for future contributors.

I've also kept in mind the contribution from @dnguneratne when sketching these wireframes out. The implementation submitted is excellent as it also helps set us up with some cleaner code which is what is being asked for here. There are a few tweaks I will suggest in #20 that will help as well.

I think our current README does not reflect the scope this project has taken. I've created a [requirements wiki](https://github.com/Mera-Gangapersaud/Creative-Collab/wiki/Scope-and-Vision) as was suggested to address the vision and scope which we can refine and edit as we go.  At the bottom is also a concerns section that I think should be kept in mind.

As for having both the back-end and front-end, the end goal was to have two different repositories but it was not reasonable with what code we had. The PR sets us up for that in the future.

Let me know what you think of the prototypes and the vison wiki while I work on reviewing the PR!
